File: src/common/py/api/project/project_events.py
import dataclasses
import typing

from ...core.messaging import (
    Message,
    Event,
)
from ...core.messaging.composers import (
    MessageBuilder,
    EventComposer,
)
from ...data.entities.connector import ConnectorInstanceID
from ...data.entities.project import Project, ProjectExternalState, ProjectID
from ...data.entities.user import UserID, UserToken


@Message.define("event/project/list")
class ProjectsListEvent(Event):
    """
    Emitted whenever the user's projects list has been updated.

    Args:
        projects: List of all projects.
    """

    projects: typing.List[Project] = dataclasses.field(default_factory=list)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        projects: typing.List[Project],
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(ProjectsListEvent, chain, projects=projects)


@Message.define("event/project/touch")
class ProjectTouchEvent(Event):
    """
    Emitted whenever a project has been "touched" (i.e., selected/activated) by the user.

    Args:
        project_id: The project ID.
    """

    project_id: ProjectID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ProjectExternalStateRenewalEvent,
            chain,
            project_id=project_id,
        )


@Message.define("event/project/logbook")
class ProjectLogbookEvent(Event):
    """
    Emitted whenever the project's logbook has been updated.

    Args:
        project_id: The project ID.
        logbook: The new project logbook.
    """

    project_id: ProjectID

    logbook: Project.Logbook

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        logbook: Project.Logbook,
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ProjectLogbookEvent, chain, project_id=project_id, logbook=logbook
        )


@Message.define("event/project/external-state")
class ProjectExternalStateEvent(Event):
    """
    Emitted whenever the external state of a project has been updated.

    Args:
        project_id: The project ID.
        user_id: The user ID.
        connector_instance: The connector instance ID.
        external_state: The new project's external state.
    """

    project_id: ProjectID
    user_id: UserID
    connector_instance: ConnectorInstanceID

    external_state: ProjectExternalState

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        user_id: UserID,
        connector_instance: ConnectorInstanceID,
        external_state: ProjectExternalState,
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ProjectExternalStateEvent,
            chain,
            project_id=project_id,
            user_id=user_id,
            connector_instance=connector_instance,
            external_state=external_state,
        )


@Message.define("event/project/external-state/renewal")
class ProjectExternalStateRenewalEvent(Event):
    """
    Emitted whenever the external state of a project needs to be updated.
    Unlike `ProjectExternalStateEvent`, the entire project needs to be sent with this request, as more than the project ID will be needed.

    Args:
        project: The project.
        connector_instance: The connector instance ID.
        user_token: The user token.
    """

    project: Project
    connector_instance: ConnectorInstanceID

    user_token: UserToken

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project: Project,
        connector_instance: ConnectorInstanceID,
        user_token: UserToken,
        chain: Message | None = None
    ) -> EventComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_event(
            ProjectExternalStateRenewalEvent,
            chain,
            project=project,
            connector_instance=connector_instance,
            user_token=user_token,
        )

File: src/common/web/api/project/ProjectEvents.ts
import { Type } from "class-transformer";

import { EventComposer } from "../../core/messaging/composers/EventComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Event } from "../../core/messaging/Event";
import { Message } from "../../core/messaging/Message";
import { type ConnectorInstanceID } from "../../data/entities/connector/ConnectorInstance";
import { ProjectLogbook } from "../../data/entities/project/logbook/ProjectLogbook";
import { Project, type ProjectID } from "../../data/entities/project/Project";
import { ProjectExternalState, ProjectUploadState } from "../../data/entities/project/ProjectExternalState";
import { type UserID } from "../../data/entities/user/User";

/**
 * Emitted whenever the user's projects list has been updated.
 *
 * @param projects - The projects list.
 */
@Message.define("event/project/list")
export class ProjectsListEvent extends Event {
    // @ts-ignore
    @Type(() => Project)
    public readonly projects: Project[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, projects: Project[], chain: Message | null = null): EventComposer<ProjectsListEvent> {
        return messageBuilder.buildEvent(ProjectsListEvent, { projects: projects }, chain);
    }
}

/**
 * Emitted whenever a project has been "touched" (i.e., selected/activated) by the user.
 *
 * @param project_id - The project ID.
 */
@Message.define("event/project/touch")
export class ProjectTouchEvent extends Event {
    public readonly project_id: ProjectID = 0;

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, projectID: ProjectID, chain: Message | null = null): EventComposer<ProjectTouchEvent> {
        return messageBuilder.buildEvent(ProjectTouchEvent, { project_id: projectID }, chain);
    }
}

/**
 * Emitted whenever the project's logbook has been updated.
 *
 * @param project_id - The project ID.
 * @param logbook - The new project logbook.
 */
@Message.define("event/project/logbook")
export class ProjectLogbookEvent extends Event {
    public readonly project_id: ProjectID = 0;

    // @ts-ignore
    @Type(() => ProjectLogbook)
    public readonly logbook: ProjectLogbook = new ProjectLogbook();

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        projectID: ProjectID,
        logbook: ProjectLogbook,
        chain: Message | null = null
    ): EventComposer<ProjectLogbookEvent> {
        return messageBuilder.buildEvent(ProjectLogbookEvent, { project_id: projectID, logbook: logbook }, chain);
    }
}

/**
 * Emitted whenever the external state of a project has been updated.
 *
 * @param project_id - The project ID.
 * @param user_id - The user ID.
 * @param connector_instance - The connector instance ID.
 * @param external_state - The new project's external state.
 */
@Message.define("event/project/external-state")
export class ProjectExternalStateEvent extends Event {
    public readonly project_id: ProjectID = 0;
    public readonly user_id: UserID = "";
    public readonly connector_instance: ConnectorInstanceID = "";

    // @ts-ignore
    @Type(() => ProjectExternalState)
    public readonly external_state: ProjectExternalState = new ProjectExternalState(ProjectUploadState.Unknown, "");

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        userID: UserID,
        projectID: ProjectID,
        connector_instance: ConnectorInstanceID,
        external_state: ProjectExternalState,
        chain: Message | null = null
    ): EventComposer<ProjectLogbookEvent> {
        return messageBuilder.buildEvent(
            ProjectLogbookEvent,
            { project_id: projectID, user_id: userID, connector_instance: connector_instance, external_state: external_state },
            chain
        );
    }
}
